[tflite] add GPU Delegate to label_image on Android

freedomtan opened the pull request:
`label_image -g 1 -m your_floating_point_model` will be delegated to GL GPU delegate on Android

Commit:
add warmup runs

freedomtan committed:
add an option to use GL ES GPU Delegate to label_image on Android

freedomtan committed:
add suport for new nnapi delegate

freedomtan committed:
clean up and fix for non-Android platforms

gbaned commented:
@freedomtan Any update please ?

freedomtan commented:
@gbaned what kind of updates?

gbaned commented:
> @gbaned what kind of updates?

@freedomtan my apologies.

gbaned commented:
@miaout17 Hi, Could you PTAL and approve.

gbaned commented:
@freedomtan could you please resolve the conflicts? Thanks!

freedomtan commented:
@gbaned rebased.

tensorflowbutler commented:
Nagging Reviewer @miaout17, @jianlijianli: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

miaout17 commented:
Sorry for the delay. Load-balancing to @srjoglekar246 -- could you review delegate-related changes?

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -112,7 +160,8 @@ void RunInference(Settings* s) {
   }
 
-  interpreter->UseNNAPI(s->accel);
+  // interpreter->UseNNAPI(s->accel);
```
Remove commented line

freedomtan replied:
oops, thanks for catching this

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.h:
```diff
@@ -17,25 +17,33 @@ limitations under the License.
 
 #include "tensorflow/lite/string.h"
+#if defined(ANDROID) || defined(__ANDROID__)
+#include "tensorflow/lite/delegates/gpu/gl_delegate.h"
```
Can this import be removed?

freedomtan replied:
`#include "tensorflow/lite/delegates/gpu/gl_delegate.h"`? I guess not, need it for something like 'TfLiteGpuDelegateOptions'

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -13,40 +13,87 @@ See the License for the specific language governing permissions and
 
-#include "tensorflow/lite/examples/label_image/bitmap_helpers.h"
-#include "tensorflow/lite/examples/label_image/get_top_n.h"
+#include <fcntl.h>      // NOLINT(build/include_order)
```
Why shift these down?

freedomtan replied:
separate `#include <>`s and `#include ""`s

srjoglekar246 replied:
I see, though the convention followed in the rest of the codebase is:
1. `<xyz.h>` includes
2.  `<>` includes
3. `""` includes
...with a newline between each block.
Could you restructure it that way?

freedomtan replied:
done

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -13,40 +13,87 @@ See the License for the specific language governing permissions and
+using TfLiteDelegatePtrMap = std::map<std::string, TfLiteDelegatePtr>;
+
+#if defined(__ANDROID__)
+Interpreter::TfLiteDelegatePtr CreateGPUDelegate(Settings* s) {
```
Instead of this, can we add a variant of the function [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/evaluation/utils.cc#L93) to accept Settings?

freedomtan replied:
sounds good. I guess I can add `Interpreter::TfLiteDelegatePtr CreateGPUDelegate(tflite::FlatBufferModel* model, tflite::TfLiteGpuDelegateOptions options)`

srjoglekar246 replied:
Thanks for this!

freedomtan replied:
It turns out this could be a problem. `TfLiteGpuDelegateOptions` is a delegate specific option. With this, compiling on non-Android platforms is messy.

freedomtan committed:
addressing review issues

freedomtan committed:
removed a stupid leftover and fix header order

freedomtan committed:
make orders of including headers more like others

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -47,6 +50,51 @@ namespace label_image {
+  return tflite::evaluation::CreateGPUDelegate(s->model, &options);
+}
+
+Interpreter::TfLiteDelegatePtr CreateNNAPIDelegate() {
```
NNAPI Delegate can also be obtained from `tflite::evaluation`

freedomtan committed:
use evaluation::CreateNNAPIDelegate()

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -47,6 +50,44 @@ namespace label_image {
+TfLiteDelegatePtrMap GetDelegates(Settings* s) {
+  TfLiteDelegatePtrMap delegates;
+  if (s->gl_backend) {
+#if defined(__ANDROID__)
```
We can remove these `#if defined(__ANDROID__)` to make the code cleaner. We can just check if returned value from `CreateGPUDelegate(s)` or `CreateNNAPIDelegate()` was nullptr using `if (!delegate)` and just avoid populating it into the delegates map.

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -153,6 +195,18 @@ void RunInference(Settings* s) {
     LOG(INFO) << "number of outputs: " << outputs.size() << "\n";
   }
 
+#if defined(__ANDROID__)
```
Please see the other comment about removing these `if` blocks. If the platform isn't Android and `GetDelegates` returns an empty map, the code block below should still work.

srjoglekar246 reviewed (changes requested):
Last couple of changes, then this should be ready :-)

freedomtan committed:
remove ifdef for delegates

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -13,6 +13,9 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "tensorflow/lite/examples/label_image/bitmap_helpers.h"
```
nit: include ordering

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -47,6 +50,43 @@ namespace label_image {
+    auto delegate = CreateGPUDelegate(s);
+    if (!delegate)
+      LOG(INFO) << "GPU acceleration is unsupported on this platform.";
+    else
```
Braces around if/else statements? Here and elsewhere in this file.

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -153,6 +194,18 @@ void RunInference(Settings* s) {
     LOG(INFO) << "number of outputs: " << outputs.size() << "\n";
   }
 
+#if defined(__ANDROID__)
```
We can remove this `#if defined(__ANDROID__)`

freedomtan committed:
addressing review comments

unfortunately, without #ifdef __ANDROID__, this won't compile yet.
need a good way to deal with 'TfLiteGpuDelegateOptions' on non-
Android platforms

freedomtan committed:
make it compile on non-Android platforms

freedomtan commented:
addressing last comments, fixed non-Android platform problem, rebased.

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -26,19 +28,22 @@ limitations under the License.
 #include "tensorflow/lite/profiling/profiler.h"
 #include "tensorflow/lite/string_util.h"
+#include "tensorflow/lite/tools/evaluation/utils.h"
+
```
Please remove newline. The headers should typically be together in alphabetical order.

srjoglekar246 reviewed (commented):
Looks good apart from one last nit.
Also, (if you haven't already), please confirm that the code runs as expected (ideally by pasting the logs/output). I will approve in the next review.
Thanks for these changes!

freedomtan committed:
remove a new line to be more adhere to the coding style

srjoglekar246 commented on tensorflow/lite/examples/label_image/label_image.cc:
```diff
@@ -26,19 +28,21 @@ limitations under the License.
 #include "tensorflow/lite/profiling/profiler.h"
 #include "tensorflow/lite/string_util.h"
+#include "tensorflow/lite/tools/evaluation/utils.h"
+#include "tensorflow/lite/examples/label_image/bitmap_helpers.h"
```
`examples/` after `delegates/` & before `kernels/`

freedomtan replied:
These two headers are in label_image. I think they are "7. Your project's .h files" described the [coding style](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)". Do I read it right?

freedomtan committed:
Merge branch 'master' into label_image_android_gpu_delegate

freedomtan commented:
merge to resolve conflicts with https://github.com/tensorflow/tensorflow/pull/30090
